src/application/services/authentication/database-user-authentication.test.js

```diff
@@ -0,0 +1,185 @@
+
+    it('must throw an error if retrieveByEmail throws', async () => {
+      const { sut, repositoryStub } = makeSut();
+      jest
+        .spyOn(repositoryStub, 'retrieveByEmail')
+        .mockReturnValueOnce(
+          new Promise((resolve, reject) => reject(new Error()))
+        );
+      const result = sut.authenticate(makeAuthenticationData());
+      await expect(result).rejects.toThrow();
+    });
+  });
+
+  describe('HashComparerService', () => {
+    it('must call compare with correct data', async () => {
+      const { sut, hashServiceStub } = makeSut();
+      const compareSpy = jest.spyOn(hashServiceStub, 'compare');
+      await sut.authenticate(makeAuthenticationData());
+      expect(compareSpy).toHaveBeenCalledWith(
+        'any_password',
+        'hashed_password'
+      );
+    });
+
+    it('must return null if compare returns false', async () => {
+      const { sut, hashServiceStub } = makeSut();
+      jest
+        .spyOn(hashServiceStub, 'compare')
+        .mockReturnValueOnce(new Promise((resolve) => resolve(false)));
+      const accessToken = await sut.authenticate(makeAuthenticationData());
+      expect(accessToken).toBeNull();
+    });
+
+    it('must throw an error if compare throws', async () => {
+      const { sut, hashServiceStub } = makeSut();
+      jest
+        .spyOn(hashServiceStub, 'compare')
+        .mockReturnValueOnce(
+          new Promise((resolve, reject) => reject(new Error()))
+        );
+      const result = sut.authenticate(makeAuthenticationData());
+      await expect(result).rejects.toThrow();
+    });
+  });
+
+  describe('TokenGeneratorService', () => {
+    it('must call generate with correct values', async () => {
+      const { sut, tokenGeneratorServiceStub } = makeSut();
+      const generateSpy = jest.spyOn(tokenGeneratorServiceStub, 'generate');
+      await sut.authenticate(makeAuthenticationData());
+      expect(generateSpy).toHaveBeenCalledWith('an_id');
+    });
+
+    it('must throw if generate throws', async () => {
+      const { sut, tokenGeneratorServiceStub } = makeSut();
+      jest
+        .spyOn(tokenGeneratorServiceStub, 'generate')
+        .mockReturnValueOnce(
+          new Promise((resolve, reject) => reject(new Error()))
+        );
+      const result = sut.authenticate(makeAuthenticationData());
+      await expect(result).rejects.toThrow();
+    });
+  });
+
+  describe('update Repository', () => {
+    it('must call updateAcessToken with correct data', async () => {
+      const { sut, repositoryStub } = makeSut();
+      const updateAcessTokenSpy = jest.spyOn(
+        repositoryStub,
+        'updateAccessToken'
+      );
+      await sut.authenticate(makeAuthenticationData());
+      expect(updateAcessTokenSpy).toHaveBeenCalledWith('an_id', 'any_token');
+    });
+
+    it('must throw if updateAcessToken throws', async () => {
+      const { sut, repositoryStub } = makeSut();
+      jest
+        .spyOn(repositoryStub, 'updateAccessToken')
```

Comment:
Pra que serve esse spyOn?

Reply:
O spyOn é como se fosse um observador, ele mocka a função passada comostring e fica observando se o método dentro da string vai ser executado. Se for ele permite que vc acesse as informações de parâmetros passados, quantas vezes foi chamado, etc.

Reply:
Ele permite que vc faça um mock da implementação da função e do valor de retorno também caso precise